src/runtime_src/core/tools/xbmgmt2/OO_Hotplug.cpp

```diff
@@ -0,0 +1,162 @@
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  if(m_help) {
+    printHelp();
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+  // Exit if neither action or device specified
+  if(m_action.empty()) {
+    printHelp();
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+
+  try {
+    if (!boost::iequals(m_action, "online") && !boost::iequals(m_action, "offline")) {
+      std::cerr << boost::format("ERROR: Invalid action value: '%s'\n") % m_action;
+      printHelp();
+      throw xrt_core::error(std::errc::operation_canceled);
+    }
+    bool is_online = boost::iequals(m_action, "online");
+
+    // Device BDF need to spcify for offline (hot removal) case
+    if (!is_online) {
+      if(m_devices.empty()) {
+        std::cerr << boost::format("ERROR: A device needs to be specified.\n");
+        throw xrt_core::error(std::errc::operation_canceled);
+      }
+    }
+    else {
+      if(!m_devices.empty()) {
+        std::cerr << boost::format("ERROR: Please do not specify any device for online.\n");
+        throw xrt_core::error(std::errc::operation_canceled);
+      }
+    }
+
+    // Collect all the devices of interest
+    std::set<std::string> deviceNames;
+    xrt_core::device_collection deviceCollection;
+    for (const auto & deviceName : m_devices) 
+      deviceNames.insert(boost::algorithm::to_lower_copy(deviceName));
+    
+    XBUtilities::collect_devices(deviceNames, false /*inUserDomain*/, deviceCollection);
+
+    // enforce 1 device specification
+    if(deviceCollection.size() > 1) {
+      std::stringstream errmsg;
+      errmsg << "Multiple devices are not supported. Please specify a single device using --device option\n\n";
+      errmsg << "List of available devices:\n";
+      boost::property_tree::ptree available_devices = XBUtilities::get_available_devices(true);
+      for(auto& kd : available_devices) {
+        boost::property_tree::ptree& _dev = kd.second;
+        errmsg << boost::format("  [%s] : %s\n") % _dev.get<std::string>("bdf") % _dev.get<std::string>("vbnv");
+      }
+      throw xrt_core::error(std::errc::operation_canceled, errmsg.str());
+    }
+
+    XBUtilities::sudo_or_throw("Root privileges required to perform hotplug operation");
+    std::cout << "CAUTION: Performing hotplug command. " <<
+	  "This command is going to impact both user pf and mgmt pf.\n" <<
+	  "Please make sure no application is currently running." << std::endl;
+
+    // Get permission from user.
+    if(!XBUtilities::can_proceed(XBUtilities::getForce()))
+      throw xrt_core::error(std::errc::operation_canceled);
+
+    if (is_online) {
+      // For Online we don't need any specific device. We are passing first device 
+      // just for accessing the sysfs entry i.e. /sys/bus/pci/rescan 
+      auto dev = xrt_core::get_mgmtpf_device(0);
+      xrt_core::device_query<xrt_core::query::hotplug_online>(dev);
+    }
+    else {
+      auto &dev = deviceCollection[0];
```

Comment:
offline doesnt require device right? why are we taking the device again? to call query?. This looks tricky.

Reply:
Offline needs a device to remove.